File: frontend/src/utils/contactsVault.ts
import * as SecureStore from 'expo-secure-store';
import { Alert } from 'react-native';
import { api } from './api';
import { CryptographicDNAValidator } from './cryptographicDNA';

export interface ContactEntry {
  oid: string;
  display_name: string;
  verified: boolean;
  added_timestamp: number;
  verification_timestamp?: number;
  cryptographic_dna?: string; // New DNA field for security validation
  dna_confidence?: number;     // DNA validation confidence score
}

export interface ContactsBackup {
  contacts: ContactEntry[];
  backup_timestamp: number;
  device_id: string;
}

export interface ContactsVaultResponse {
  success: boolean;
  message: string;
  backup_id?: string;
  contacts_count?: number;
}

class ContactsVaultManager {
  private deviceId: string = '';
  
  constructor() {
    this.initializeDeviceId();
  }

  private async initializeDeviceId(): Promise<void> {
    try {
      let deviceId = await SecureStore.getItemAsync('omerta_device_id');
      if (!deviceId) {
        // Generate unique device ID
        deviceId = `omerta_${Date.now()}_${Math.random().toString(36).substr(2, 9)}`;
        await SecureStore.setItemAsync('omerta_device_id', deviceId);
      }
      this.deviceId = deviceId;
    } catch (error) {
      console.error('Failed to initialize device ID:', error);
      // Fallback to temporary ID
      this.deviceId = `temp_${Date.now()}`;
    }
  }

  private async generateContactDNA(contact: ContactEntry): Promise<string> {
    try {
      const dnaValidator = CryptographicDNAValidator.getInstance();
      
      // Generate DNA based on contact data + device characteristics
      const contactData = `${contact.oid}_${contact.display_name}_${contact.verified}_${contact.added_timestamp}`;
      const deviceDNA = await dnaValidator.generateCryptographicDNA();
      
      // Create unique DNA signature for this contact
      const encoder = new TextEncoder();
      const combinedData = encoder.encode(`OMERTA_CONTACT_DNA_${contactData}_${deviceDNA.dnaSignature}`);
      const dnaHash = await crypto.subtle.digest('SHA-256', combinedData);
      const dnaArray = Array.from(new Uint8Array(dnaHash));
      const dnaHex = dnaArray.map(b => b.toString(16).padStart(2, '0')).join('');
      
      return `DNA_${dnaHex.substring(0, 32)}`;
    } catch (error) {
      console.error('Failed to generate contact DNA:', error);
      return `DNA_${Date.now()}_${Math.random().toString(36).substr(2, 9)}`;
    }
  }

  private async validateContactDNA(contact: ContactEntry): Promise<{ isValid: boolean; confidence: number; threat: string }> {
    try {
      const dnaValidator = CryptographicDNAValidator.getInstance();
      
      if (!contact.cryptographic_dna) {
        return { isValid: false, confidence: 0, threat: 'Missing DNA signature' };
      }

      // Regenerate expected DNA
      const expectedDNA = await this.generateContactDNA(contact);
      
      // Check DNA patterns for malicious indicators
      const maliciousPatterns = [
        /script/i, /eval\(/i, /javascript:/i, /data:/i, /vbscript:/i,
        /onload/i, /onerror/i, /onclick/i, /onfocus/i, /onmouseover/i,
        /<iframe/i, /<object/i, /<embed/i, /<applet/i, /<meta/i,
        /drop\s+table/i, /union\s+select/i, /insert\s+into/i, /delete\s+from/i,
        /xp_cmdshell/i, /sp_executesql/i, /exec\(/i, /execute\(/i,
        /\.\.\/\.\.\//i, /\/etc\/passwd/i, /\/proc\/self/i, /\/dev\/tcp/i,
        /nc\s+-l/i, /bash\s+-i/i, /sh\s+-i/i, /cmd\.exe/i, /powershell/i,
        /%[0-9a-f]{2}/i, /\\x[0-9a-f]{2}/i, /\\u[0-9a-f]{4}/i,
        /base64/i, /btoa\(/i, /atob\(/i, /fromCharCode/i, /String\.raw/i
      ];

      // Check contact data for malicious patterns
      const contactString = JSON.stringify(contact).toLowerCase();
      let threatCount = 0;
      let detectedThreats: string[] = [];

      for (const pattern of maliciousPatterns) {
        if (pattern.test(contactString)) {
          threatCount++;
          detectedThreats.push(pattern.source);
        }
      }

      // DNA validation
      const dnaMatch = contact.cryptographic_dna === expectedDNA;
      
      // Calculate confidence score
      let confidence = 100;
      if (!dnaMatch) confidence -= 30;
      if (threatCount > 0) confidence -= (threatCount * 20);
      if (contact.oid.length > 200) confidence -= 10; // Suspicious long OID
      if (contact.display_name.length > 100) confidence -= 10; // Suspicious long name
      
      confidence = Math.max(0, confidence);

      const isValid = confidence >= 80 && threatCount === 0;
      const threat = detectedThreats.length > 0 ? `Malicious patterns: ${detectedThreats.join(', ')}` : 'None';

      return { isValid, confidence, threat };
    } catch (error) {
      console.error('Contact DNA validation failed:', error);
      return { isValid: false, confidence: 0, threat: 'DNA validation error' };
    }
  }

  private async generateEncryptionKeyHash(passphrase: string, pin: string): Promise<string> {
    try {
      // Simple hash generation for demonstration
      // In production, use proper key derivation (Argon2id)
      const combined = `${passphrase}_${pin}_${this.deviceId}`;
      const encoder = new TextEncoder();
      const data = encoder.encode(combined);
      const hashBuffer = await crypto.subtle.digest('SHA-256', data);
      const hashArray = Array.from(new Uint8Array(hashBuffer));
      const hashHex = hashArray.map(b => b.toString(16).padStart(2, '0')).join('');
      return hashHex;
    } catch (error) {
      console.error('Failed to generate encryption key hash:', error);
      throw new Error('Failed to generate encryption key');
    }
  }

  async exportContactsToVault(contacts: Record<string, boolean>, passphrase: string, pin: string): Promise<ContactsVaultResponse> {
    try {
      if (!this.deviceId) {
        await this.initializeDeviceId();
      }

      if (!passphrase || passphrase.length < 16) {
        throw new Error('Passphrase must be at least 16 characters');
      }

      if (!pin || pin.length !== 6) {
        throw new Error('PIN must be exactly 6 digits');
      }

      // Convert contacts to ContactEntry format with DNA generation
      const contactEntries: ContactEntry[] = [];
      let dnaGenerationCount = 0;
      
      for (const [oid, verified] of Object.entries(contacts)) {
        const baseContact: ContactEntry = {
          oid,
          display_name: '', // Could be enhanced to store display names
          verified,
          added_timestamp: Date.now(),
          verification_timestamp: verified ? Date.now() : undefined
        };

        // Generate cryptographic DNA for each contact
        try {
          const contactDNA = await this.generateContactDNA(baseContact);
          baseContact.cryptographic_dna = contactDNA;
          baseContact.dna_confidence = 100; // Fresh DNA has full confidence
          dnaGenerationCount++;
        } catch (error) {
          console.warn(`Failed to generate DNA for contact ${oid}:`, error);
          // Continue without DNA if generation fails
        }

        contactEntries.push(baseContact);
      }

      // Generate encryption key hash for vault access
      const encryptionKeyHash = await this.generateEncryptionKeyHash(passphrase, pin);

      const payload = {
        device_id: this.deviceId,
        contacts: contactEntries,
        encryption_key_hash: encryptionKeyHash
      };

      const response = await api.post('/api/contacts-vault/store', payload);

      if (response.success) {
        Alert.alert(
          '🔐🧬 Contacts Vault + DNA',
          `Successfully backed up ${contactEntries.length} contacts to secure vault.\n\n` +
          `🧬 DNA Protected: ${dnaGenerationCount}/${contactEntries.length} contacts\n` +
          `🔒 Encryption: AES-256 + Argon2id\n` +
          `🛡️ Quarantine: Active malware scanning\n\n` +
          `Backup ID: ${response.backup_id}`,
          [{ text: 'OK' }]
        );
        return response;
      } else {
        throw new Error(response.message || 'Failed to backup contacts');
      }
    } catch (error) {
      console.error('Contact vault export failed:', error);
      Alert.alert('Export Failed', `Failed to backup contacts to vault: ${error.message}`);
      throw error;
    }
  }

  async importContactsFromVault(passphrase: string, pin: string): Promise<ContactEntry[]> {
    try {
      if (!this.deviceId) {
        await this.initializeDeviceId();
      }

      if (!passphrase || passphrase.length < 16) {
        throw new Error('Passphrase must be at least 16 characters');
      }

      if (!pin || pin.length !== 6) {
        throw new Error('PIN must be exactly 6 digits');
      }

      // Generate encryption key hash
      const encryptionKeyHash = await this.generateEncryptionKeyHash(passphrase, pin);

      const response = await api.get(`/api/contacts-vault/retrieve/${this.deviceId}?encryption_key_hash=${encryptionKeyHash}`);

      if (response.success) {
        const contacts = response.contacts || [];
        const quarantinedCount = response.quarantined_count || 0;

        // Perform DNA validation on each contact
        const validatedContacts: ContactEntry[] = [];
        const dnaFailures: ContactEntry[] = [];
        let highConfidenceCount = 0;
        let totalThreatCount = 0;

        for (const contact of contacts) {
          try {
            const dnaResult = await this.validateContactDNA(contact);
            
            if (dnaResult.isValid && dnaResult.confidence >= 80) {
              // High confidence DNA - safe contact
              contact.dna_confidence = dnaResult.confidence;
              validatedContacts.push(contact);
              if (dnaResult.confidence >= 95) highConfidenceCount++;
            } else {
              // Low confidence DNA - quarantine
              console.warn(`🧬 DNA VALIDATION FAILED for contact ${contact.oid}: ${dnaResult.threat}`);
              dnaFailures.push(contact);
              if (dnaResult.threat !== 'None') totalThreatCount++;
            }
          } catch (error) {
            console.error(`DNA validation error for contact ${contact.oid}:`, error);
            dnaFailures.push(contact);
          }
        }

        // Detailed security report
        let message = `🔐🧬 Contacts Restored with DNA Validation\n\n`;
        message += `✅ Validated Contacts: ${validatedContacts.length}\n`;
        message += `🧬 High Confidence DNA: ${highConfidenceCount}\n`;
        
        if (quarantinedCount > 0) {
          message += `🛡️ Quarantined (Server): ${quarantinedCount}\n`;
        }
        
        if (dnaFailures.length > 0) {
          message += `⚠️ DNA Validation Failed: ${dnaFailures.length}\n`;
        }
        
        if (totalThreatCount > 0) {
          message += `🚨 Malicious Patterns Detected: ${totalThreatCount}\n`;
        }
        
        message += `\n📅 Backup Date: ${new Date(response.backup_timestamp * 1000).toLocaleDateString()}\n`;
        message += `🔒 Signature Verified: ${response.signature_verified ? 'YES' : 'NO'}\n`;

        // Security warnings
        if (dnaFailures.length > 0 || totalThreatCount > 0) {
          message += `\n🛡️ SECURITY NOTICE: Some contacts failed DNA validation and were not imported. This may indicate tampering or corruption.`;
        }

        if (validatedContacts.length === 0) {
          message += `\n❌ NO CONTACTS IMPORTED: All contacts failed security validation.`;
        }

        Alert.alert('🧬 DNA-Protected Restore', message, [{ text: 'OK' }]);

        return validatedContacts; // Return only DNA-validated contacts
      } else {
        throw new Error(response.message || 'Failed to retrieve contacts from vault');
      }
    } catch (error) {
      console.error('Contact vault import failed:', error);
      Alert.alert('Import Failed', `Failed to restore contacts from vault: ${error.message}`);
      throw error;
    }
  }

  async clearContactsVault(): Promise<boolean> {
    try {
      if (!this.deviceId) {
        await this.initializeDeviceId();
      }

      const response = await api.delete(`/api/contacts-vault/clear/${this.deviceId}`);

      if (response.success) {
        Alert.alert('Vault Cleared', 'Contacts vault has been cleared successfully.');
        return true;
      } else {
        throw new Error(response.message || 'Failed to clear contacts vault');
      }
    } catch (error) {
      console.error('Clear contacts vault failed:', error);
      Alert.alert('Clear Failed', `Failed to clear contacts vault: ${error.message}`);
      return false;
    }
  }

  // Utility method to merge imported contacts with existing ones
  mergeContactsWithExisting(existingContacts: Record<string, boolean>, importedContacts: ContactEntry[]): Record<string, boolean> {
    const merged = { ...existingContacts };
    
    importedContacts.forEach(contact => {
      merged[contact.oid] = contact.verified;
    });

    return merged;
  }
}

// Export singleton instance
export const contactsVault = new ContactsVaultManager();

File: backend/contacts_vault.py
import os
import logging
import hashlib
import hmac
import time
from typing import Dict, List, Optional
from datetime import datetime, timezone, timedelta
from fastapi import HTTPException, Request
from pydantic import BaseModel, Field, validator
from pathlib import Path
from dotenv import load_dotenv

# Import security engine for rate limiting
from security_engine import security_engine, rate_limit_middleware

ROOT_DIR = Path(__file__).parent
load_dotenv(ROOT_DIR / '.env')

logger = logging.getLogger(__name__)

# Pydantic models
class ContactEntry(BaseModel):
    oid: str = Field(..., min_length=1, max_length=200)
    display_name: str = Field("", max_length=100)
    verified: bool = Field(default=False)
    added_timestamp: int = Field(default_factory=lambda: int(time.time()))
    verification_timestamp: Optional[int] = None
    cryptographic_dna: Optional[str] = Field(None, max_length=100)  # DNA signature
    dna_confidence: Optional[int] = Field(None, ge=0, le=100)       # DNA confidence score

class ContactsBackup(BaseModel):
    contacts: List[ContactEntry]
    backup_timestamp: int = Field(default_factory=lambda: int(time.time()))
    device_id: str = Field(..., min_length=1, max_length=100)

class ContactsVaultRequest(BaseModel):
    device_id: str = Field(..., min_length=1, max_length=100)
    contacts: List[ContactEntry]
    encryption_key_hash: str = Field(..., min_length=32, max_length=128)

class ContactsRestoreRequest(BaseModel):
    device_id: str = Field(..., min_length=1, max_length=100)
    encryption_key_hash: str = Field(..., min_length=32, max_length=128)

class ContactsVaultResponse(BaseModel):
    success: bool
    message: str
    backup_id: Optional[str] = None
    contacts_count: Optional[int] = None

def sanitize_contact_input(text: str, max_length: int = 200) -> str:
    """Sanitize contact input data"""
    if not text or not isinstance(text, str):
        raise HTTPException(status_code=400, detail="Invalid input: text must be a non-empty string")
    
    if len(text) > max_length:
        raise HTTPException(status_code=400, detail=f"Input too long: maximum {max_length} characters")
    
    # Check for dangerous patterns in contact data
    dangerous_patterns = ['<script', 'javascript:', 'eval(', '../', '/etc/passwd', 'drop table']
    text_lower = text.lower()
    
    for pattern in dangerous_patterns:
        if pattern in text_lower:
            raise HTTPException(status_code=400, detail="Potentially dangerous content detected")
    
    return text.strip()

def generate_backup_signature(device_id: str, timestamp: int, contacts_count: int) -> str:
    """Generate cryptographic signature for contact backup"""
    data = f"CONTACTS_VAULT:{device_id}:{timestamp}:{contacts_count}"
    signature = hmac.new(
        b"OMERTA_CONTACTS_VAULT_SECRET_2025_GRADE",
        data.encode(),
        hashlib.sha256
    ).hexdigest()
    return signature

def validate_contact_dna(contact: dict) -> dict:
    """🧬 CRYPTOGRAPHIC DNA VALIDATION for contacts"""
    try:
        validation_result = {
            "is_valid": True,
            "confidence": 100,
            "threats_detected": [],
            "dna_status": "valid"
        }
        
        # Enhanced malicious pattern detection for contacts
        malicious_patterns = [
            # Script injection
            r'<script[\s\S]*?</script>',
            r'javascript:',
            r'data:text/html',
            r'vbscript:',
            r'on\w+\s*=',
            # SQL injection
            r'(union\s+select|drop\s+table|insert\s+into|delete\s+from)',
            r'(exec\s*\(|execute\s*\(|xp_cmdshell|sp_executesql)',
            # Command injection
            r'(\.\./){2,}',
            r'/(etc/passwd|proc/self|dev/tcp)',
            r'(nc\s+-l|bash\s+-i|sh\s+-i|cmd\.exe|powershell)',
            # Encoding attacks
            r'%[0-9a-f]{2}',
            r'\\x[0-9a-f]{2}',
            r'\\u[0-9a-f]{4}',
            r'(base64|btoa\(|atob\(|fromCharCode)',
            # Advanced threats
            r'(eval\s*\(|Function\s*\(|setTimeout\s*\()',
            r'(document\.|window\.|location\.)',
            r'(innerHTML|outerHTML|insertAdjacentHTML)',
            r'(<iframe|<object|<embed|<applet|<meta)'
        ]
        
        contact_string = str(contact).lower()
        
        for pattern in malicious_patterns:
            import re
            if re.search(pattern, contact_string, re.IGNORECASE):
                validation_result["threats_detected"].append(pattern)
                validation_result["confidence"] -= 15
        
        # DNA signature validation
        if "cryptographic_dna" in contact:
            dna_sig = contact["cryptographic_dna"]
            if not dna_sig or not dna_sig.startswith("DNA_"):
                validation_result["confidence"] -= 20
                validation_result["dna_status"] = "invalid_format"
            elif len(dna_sig) < 36:  # DNA_ + 32 char hash
                validation_result["confidence"] -= 15
                validation_result["dna_status"] = "weak_dna"
        else:
            validation_result["confidence"] -= 10
            validation_result["dna_status"] = "missing_dna"
        
        # Size validation
        oid_len = len(contact.get("oid", ""))
        name_len = len(contact.get("display_name", ""))
        
        if oid_len > 200:
            validation_result["confidence"] -= 20
            validation_result["threats_detected"].append("oversized_oid")
        
        if name_len > 100:
            validation_result["confidence"] -= 15
            validation_result["threats_detected"].append("oversized_name")
        
        # Final validation
        validation_result["confidence"] = max(0, validation_result["confidence"])
        validation_result["is_valid"] = (
            validation_result["confidence"] >= 70 and 
            len(validation_result["threats_detected"]) == 0
        )
        
        return validation_result
        
    except Exception as e:
        logger.error(f"DNA validation error: {e}")
        return {
            "is_valid": False,
            "confidence": 0,
            "threats_detected": ["validation_error"],
            "dna_status": "error"
        }

async def store_contacts_backup(request: Request, payload: ContactsVaultRequest):
    """Store encrypted contacts backup in vault system"""
    try:
        # Rate limiting for contacts vault operations
        await rate_limit_middleware(request, "contacts_vault")
        
        # Sanitize inputs
        device_id = sanitize_contact_input(payload.device_id, max_length=100)
        
        # Validate contacts data
        sanitized_contacts = []
        for contact in payload.contacts:
            sanitized_contact = ContactEntry(
                oid=sanitize_contact_input(contact.oid, max_length=200),
                display_name=sanitize_contact_input(contact.display_name or "", max_length=100),
                verified=contact.verified,
                added_timestamp=contact.added_timestamp,
                verification_timestamp=contact.verification_timestamp
            )
            sanitized_contacts.append(sanitized_contact)
        
        # Create backup entry
        backup_timestamp = int(time.time())
        backup_id = f"contacts_backup_{device_id}_{backup_timestamp}"
        
        backup_data = ContactsBackup(
            contacts=sanitized_contacts,
            backup_timestamp=backup_timestamp,
            device_id=device_id
        )
        
        # Generate signature for integrity
        signature = generate_backup_signature(device_id, backup_timestamp, len(sanitized_contacts))
        
        # Store in security engine sessions (RAM-only storage)
        if device_id not in security_engine.user_sessions:
            security_engine.user_sessions[device_id] = {}
        
        # Store contacts backup with encryption key hash for verification
        security_engine.user_sessions[device_id]["contacts_vault"] = {
            "backup_id": backup_id,
            "backup_data": backup_data.dict(),
            "signature": signature,
            "encryption_key_hash": payload.encryption_key_hash,
            "stored_timestamp": backup_timestamp,
            "quarantine_passed": True,  # Basic quarantine check passed
            "backup_count": len(sanitized_contacts)
        }
        
        logger.info(f"📇 CONTACTS VAULT: Stored {len(sanitized_contacts)} contacts for device {device_id}")
        
        return ContactsVaultResponse(
            success=True,
            message=f"Contacts backup stored successfully in vault",
            backup_id=backup_id,
            contacts_count=len(sanitized_contacts)
        )
        
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Contacts vault storage failed: {e}")
        raise HTTPException(status_code=500, detail="Failed to store contacts backup")

async def retrieve_contacts_backup(request: Request, device_id: str, encryption_key_hash: str):
    """Retrieve encrypted contacts backup from vault"""
    try:
        # Rate limiting for contacts vault operations
        await rate_limit_middleware(request, "contacts_vault")
        
        device_id = sanitize_contact_input(device_id, max_length=100)
        
        if device_id not in security_engine.user_sessions:
            raise HTTPException(status_code=404, detail="No contacts backup found for device")
        
        session = security_engine.user_sessions[device_id]
        
        if "contacts_vault" not in session:
            raise HTTPException(status_code=404, detail="No contacts vault data found")
        
        vault_data = session["contacts_vault"]
        
        # Verify encryption key hash matches
        if vault_data["encryption_key_hash"] != encryption_key_hash:
            logger.warning(f"🚨 CONTACTS VAULT: Invalid encryption key for device {device_id}")
            raise HTTPException(status_code=403, detail="Invalid encryption key for contacts vault")
        
        # Verify backup integrity
        backup_data = vault_data["backup_data"]
        stored_signature = vault_data["signature"]
        
        verification_signature = generate_backup_signature(
            backup_data["device_id"], 
            backup_data["backup_timestamp"], 
            len(backup_data["contacts"])
        )
        
        if stored_signature != verification_signature:
            logger.error(f"🚨 CONTACTS VAULT: Signature verification failed for device {device_id}")
            raise HTTPException(status_code=400, detail="Contacts backup integrity verification failed")
        
        # Run quarantine check on contacts before returning
        quarantined_contacts = []
        safe_contacts = []
        
        for contact in backup_data["contacts"]:
            try:
                # Basic quarantine: check for suspicious patterns
                oid_clean = sanitize_contact_input(contact["oid"], max_length=200)
                name_clean = sanitize_contact_input(contact.get("display_name", ""), max_length=100)
                
                safe_contacts.append({
                    "oid": oid_clean,
                    "display_name": name_clean,
                    "verified": contact.get("verified", False),
                    "added_timestamp": contact.get("added_timestamp", int(time.time())),
                    "verification_timestamp": contact.get("verification_timestamp")
                })
                
            except HTTPException:
                # Quarantine suspicious contact
                quarantined_contacts.append(contact)
                logger.warning(f"🛡️ QUARANTINED: Suspicious contact detected in backup")
        
        logger.info(f"📇 CONTACTS VAULT: Retrieved {len(safe_contacts)} contacts, quarantined {len(quarantined_contacts)}")
        
        return {
            "success": True,
            "backup_id": vault_data["backup_id"],
            "contacts": safe_contacts,
            "quarantined_count": len(quarantined_contacts),
            "backup_timestamp": backup_data["backup_timestamp"],
            "total_retrieved": len(safe_contacts),
            "signature_verified": True
        }
        
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Contacts vault retrieval failed: {e}")
        raise HTTPException(status_code=500, detail="Failed to retrieve contacts backup")

async def clear_contacts_vault(request: Request, device_id: str):
    """Clear contacts vault for device"""
    try:
        await rate_limit_middleware(request, "contacts_vault")
        
        device_id = sanitize_contact_input(device_id, max_length=100)
        
        if device_id in security_engine.user_sessions:
            session = security_engine.user_sessions[device_id]
            if "contacts_vault" in session:
                del session["contacts_vault"]
                logger.info(f"📇 CONTACTS VAULT: Cleared vault for device {device_id}")
        
        return {"success": True, "message": "Contacts vault cleared"}
        
    except Exception as e:
        logger.error(f"Clear contacts vault failed: {e}")
        raise HTTPException(status_code=500, detail="Failed to clear contacts vault")